vortex-expr/src/pruning/pruning_predicate.rs

```diff
@@ -67,3 +90,279 @@ impl PruningPredicate {
         Ok(Some(self.expr.evaluate(&Scope::new(metadata.to_array()))?))
     }
 }
+
+#[derive(Default)]
+struct FileStatsCatalog {
+    usage: HashMap<(Identifier, FieldPath, Stat), ExprRef>,
+}
+
+impl StatsCatalog for FileStatsCatalog {
+    fn stats_ref(&mut self, id: &Identifier, field: &FieldPath, stat: Stat) -> Option<ExprRef> {
+        let mut expr = var(id.clone());
+        let name = field
+            .path()
+            .iter()
+            .map(|f| match f {
+                Field::Name(n) => n.as_ref(),
+                Field::ElementType => todo!("element type not currently handled"),
+            })
+            .chain(iter::once(stat.name()))
+            .join("_");
+        expr = get_item(name, expr);
+        self.usage
+            .insert((id.clone(), field.clone(), stat), expr.clone());
+        Some(expr)
+    }
+}
+
+#[allow(clippy::type_complexity)]
+// TODO: remove (Id, FieldPath) when updating FieldPath
+pub fn pruning_expr(expr: &ExprRef) -> Option<(ExprRef, Relation<(Identifier, FieldPath), Stat>)> {
```

Comment:
dear god this return type

Reply:
its gettingg better

Reply:
I have a follow up that create an access path = (Identifier, FieldPath)